Split check_test_cases.py and outcome_analysis.py

gilles-peskine-arm opened the pull request:
* Split `check_test_cases.py` into test case collection (`collect_test_cases.py`, used by outcome analysis) and the checks proper. Move both parts to the framework.
* Split `analyze_outcomes.py` into code (`outcome_analysis.py`, moved to the framework) and data (`analyze_outcomes.py`, stays in its branch).
* Upgrade mypy, for the sake of a new feature in `outcome_analysis.py`.
* Prepare to require all test cases to be executed, i.e. to enforcing full coverage. But for now, keep it as a warning in this branch. To be continued in https://github.com/Mbed-TLS/mbedtls/pull/9593.

## PR checklist

- [x] **changelog** not required because: test stuff only
- [x] **development PR** here
- [x] **framework PR** provided Mbed-TLS/mbedtls-framework#55
- [x] **3.6 PR** provided #9669
- [x] **2.28 PR** provided # | not required because: no framework in 2.28, and the improvements to outcome analysis aren't worth backporting at this stage
- **tests**  provided

eleuzi01 reviewed (approved):
LGTM

mpg commented on tests/scripts/analyze_outcomes.py:
```diff
@@ -761,83 +494,5 @@ class DriverVSReference_block_cipher_dispatch(DriverVSReference):
-        traceback.print_exc()
-        sys.exit(120)
-
 if __name__ == '__main__':
```
Does it still make sense to have this `if` (as opposed to just calling `outcome_analysis.main()` unconditionally)?

gilles-peskine-arm replied:
I don't see why not. `if __name__ == '__main__':` is very idiomatic in Python.

mpg replied:
I know, it just doesn't look very useful here as we're unlikely to want to `import` this file IMO. Anyway, doesn't hurt (neither technically, nor readability since it's so common), so I'm marking this conversation as resolved.

mpg commented on scripts/ci.requirements.txt:
```diff
@@ -7,9 +7,13 @@
-mypy >= 0.971
+# Use a version of mypy that is compatible with our code base.
+# mypy <0.940 is known not to work: see commit
+#  :/Upgrade mypy to the last version supporting Python 3.6
```
Is `:/` intentional? If so, what does it mean?

gilles-peskine-arm replied:
It's a Git revision.
```
git show -s ':/Upgrade mypy to the last version supporting Python 3.6'
```

mpg replied:
Oh great, thanks!

mpg reviewed (approved):
Looks good to me. Only two minor comments.

gilles-peskine-arm commented:
I have rewritten the history a little so that there is now a single commit that removes the files moved to the framework, obtained with David's script. I've also updated the framework submodule to the latest version.

eleuzi01 reviewed (dismissed):
LGTM, thanks!

mpg reviewed (approved):
Re-approving after force-push - content is identical to what I previously approved, with the exception of the framework pointer updated to the latest version of https://github.com/Mbed-TLS/mbedtls-framework/pull/55

gilles-peskine-arm commented:
I just amended the last commit to switch the framework to the result of merging the companion PR.

mpg reviewed (approved):
LGTM after updating the framework pointer.

eleuzi01 reviewed (approved):
LGTM but the CI is unhappy

gilles-peskine-arm commented:
The CI failure looks infrastructure-related:
```
[2024-10-10T09:00:11.908Z] armclang: error: Failed to check out a license.
[2024-10-10T09:00:11.908Z] Unable to connect to the license server. Check that ARMLMD_LICENSE_FILE is set correctly, and the license server is available.
```

This might just be a network glitch, I'll try re-running the job.
